sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/DoFnReflector.java

```diff
@@ -780,175 +472,121 @@ private void readObject(java.io.ObjectInputStream in)
-     * definitions. Second, it uses the {@link AdditionalParameter} to populate the arguments to the
-     * method.
-     */
-    private StackManipulation pushDelegateField() {
-      return new StackManipulation.Compound(
-          // Push "this" reference to the stack
-          MethodVariableAccess.REFERENCE.loadOffset(0),
-          // Access the delegate field of the the invoker
-          FieldAccess.forField(field).getter());
+    @Override
+    protected boolean isRequired() {
+      return true;
     }
 
-    private StackManipulation pushArgument(
-        AdditionalParameter arg, MethodDescription instrumentedMethod) {
-      MethodDescription transform = arg.method;
-
-      return new StackManipulation.Compound(
-          // Push the ExtraContextFactory which must have been argument 2 of the instrumented method
-          MethodVariableAccess.REFERENCE.loadOffset(2),
-          // Invoke the appropriate method to produce the context argument
-          MethodInvocation.invoke(transform));
-    }
+    @Override
+    protected StackManipulation invokeTargetMethod(MethodDescription instrumentedMethod) {
+      MethodDescription targetMethod =
+          new MethodLocator.ForExplicitMethod(
+                  new MethodDescription.ForLoadedMethod(signature.targetMethod))
+              .resolve(instrumentedMethod);
 
-    private StackManipulation invokeTargetMethod(MethodDescription instrumentedMethod) {
-      MethodDescription targetMethod = new MethodLocator.ForExplicitMethod(
-          new MethodDescription.ForLoadedMethod(target)).resolve(instrumentedMethod);
-      ParameterList<?> params = targetMethod.getParameters();
-
-      // Instructions to setup the parameters for the call
-      ArrayList<StackManipulation> parameters = new ArrayList<>(args.size() + 1);
-      // 1. The first argument in the delegate method must be the context. This corresponds to
-      //    the first argument in the instrumented method, so copy that.
-      parameters.add(MethodVariableAccess.of(
-          params.get(0).getType().getSuperClass()).loadOffset(1));
-      // 2. For each of the extra arguments push the appropriate value.
-      for (AdditionalParameter arg : args) {
-        parameters.add(pushArgument(arg, instrumentedMethod));
+      // Parameters of the wrapper OldDoFn method:
+      //   DoFn.ProcessContext, ExtraContextFactory.
+      // Parameters of the wrapped DoFn method:
+      //   DoFn.ProcessContext, [BoundedWindow, InputProvider, OutputReceiver] in any order
+      ArrayList<StackManipulation> parameters = new ArrayList<>();
+      // Push the ProcessContext argument.
+      parameters.add(MethodVariableAccess.REFERENCE.loadOffset(1));
+      // Push the extra arguments in their actual order.
+      StackManipulation pushExtraContextFactory = MethodVariableAccess.REFERENCE.loadOffset(2);
+      for (int i = 1; i < signature.numExtraParameters() + 1; ++i) {
+        if (i == signature.boundedWindowParamIndex) {
```

Comment:
This if-ladder is partly why we shouldn't switch to use integer indices for these. A class or enum that can be added to a list is much better (it doesn't tie code generation as closely together with the parameters, which makes it much more extensible).